### PR TITLE
[Auth with tenant, 1]adjust auth-lib to have az style usage

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java [hunk 1]
```diff
@@ -64,6 +64,10 @@
     protected void doExecute() throws AzureExecutionException {
         // initialize library client
         az = getOrCreateAzureAppServiceClient();
+        if (az == null) {
+            getTelemetryProxy().trackEvent(INIT_FAILURE);
```
Comment (Flanker32) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java:68]: Do we need to get exception from Account (if any)?

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java [hunk 2]
```diff
@@ -50,4 +54,34 @@ public static boolean containsIgnoreCase(List<String> list, String str) {
     public static String getId(Object obj) {
         return Integer.toHexString(System.identityHashCode(obj));
     }
+
+    public static <T> T firstNonNull(T... args) {
+        for (T obj : args) {
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java:59]: `ObjectUtils.firstNonNull()`?

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java [hunk 3]
```diff
@@ -50,4 +54,34 @@ public static boolean containsIgnoreCase(List<String> list, String str) {
...
+            }
+        }
+        return null;
+    }
+
+    public static <T> T copyProperties(T to, T from) throws IllegalAccessException, InvocationTargetException {
+        Objects.requireNonNull(from);
+        Objects.requireNonNull(to);
+        for (Field field : FieldUtils.getAllFields(from.getClass())) {
+            FieldUtils.writeField(to, field.getName(), FieldUtils.readField(from, field.getName(), true), true);
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java:71]: For reference filed, we will just copy the reference, is this enough or we also need deep copy for that?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/model/AccountEntity.java [hunk 4]
```diff
@@ -15,15 +16,25 @@
...
+    private AzureEnvironment environment;
 
     private String email;
 
     private boolean selected;
 
     private boolean authenticated;
 
-    List<SubscriptionEntity> subscriptions;
+    private List<SubscriptionEntity> subscriptions;
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/model/AccountEntity.java:29]: As we have `List<SubscriptionEntity> subscriptions` and SubscriptionEntity will record whether it is selected, do we need to keep `selectedSubscriptions` and `selectedSubscriptionIds`?

File: azure-spring-cloud-maven-plugin/src/main/java/com/microsoft/azure/maven/springcloud/AbstractMojoBase.java [hunk 5]
```diff
@@ -146,26 +149,29 @@ public void execute() throws MojoFailureException {
...
         // Init telemetries
         initTelemetry();
+        Azure.az().config().setLogLevel(LogLevel.NONE);
+        Azure.az().config().setUserAgent(getUserAgent());
         trackMojoExecution(MojoStatus.Start);
         final MavenAuthConfiguration mavenAuthConfiguration = auth == null ? new MavenAuthConfiguration() : auth;
         mavenAuthConfiguration.setType(getAuthType());
-        this.azureCredentialWrapper = MavenAuthManager.getInstance().login(session, settingsDecrypter, mavenAuthConfiguration);
-        if (Objects.isNull(azureCredentialWrapper)) {
+        MavenAuthManager.getInstance().login(MavenAuthManager.getInstance().buildAuthConfiguration(session, settingsDecrypter, mavenAuthConfiguration));
```
Comment (Flanker32) [inline azure-spring-cloud-maven-plugin/src/main/java/com/microsoft/azure/maven/springcloud/AbstractMojoBase.java:163]: Seems that `MavenAuthManager.login` just call the `Azure.az(AzureAccount.class).login`, so shall we using `MavenAuthManager` to get the auth configuration and call az.login here?

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java (hunk 3 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java:71]: No need, just first level of copying

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/model/AccountEntity.java (hunk 4 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/model/AccountEntity.java:29]: selectedSubscriptionIds are only used to pass in selected ids, it should never called after initialized tenants and subscriptions, I will remove the getter for selectedSubscriptionIds

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java (hunk 1 above)
Comment (andxu) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java:68]: Will fix in later pr

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java (hunk 2 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/utils/Utils.java:59]: Will fix later

File: azure-spring-cloud-maven-plugin/src/main/java/com/microsoft/azure/maven/springcloud/AbstractMojoBase.java (hunk 5 above)
Comment (andxu) [reply azure-spring-cloud-maven-plugin/src/main/java/com/microsoft/azure/maven/springcloud/AbstractMojoBase.java]: The design expects you to login first and then you can get the credential anywhere

**Pull request closed without merging**